File: cpp/src/advanced/template_inheritance.cpp
#include "template_inheritance.h"


template <typename T, typename U>
Pair<T, U>::Pair(T first, U second) 
    : first(first), second(second) {;}

template <typename T, typename U>
void Pair<T, U>::set_value(T f, U s) {  first = f; second = s; } // setter
template <typename T, typename U>
T Pair<T, U>::get_first() const { return first; } // getter
template <typename T, typename U>
U Pair<T, U>::get_second() const { return second; }


template <typename T, typename U>
LabeledPair<T, U>::LabeledPair(T first, U second, string label)
    : Pair<T, U>(first, second), label(label) {;}
template <typename T, typename U>
void LabeledPair<T, U>::stampa_label() const { cout << "Label: " << label << endl;  }

void inheritance_template_example () {
  Pair<int, double> p(10, 20.5); // Creazione di un oggetto Pair
  cout << "First: " << p.get_first() << ", Second: " << p.get_second() << endl;

  LabeledPair<int, double> lp(30, 40.5, "Example Label"); // Creazione di un oggetto LabeledPair
  cout << "First: " << lp.get_first() << ", Second: " << lp.get_second() << endl;
  lp.stampa_label();
}

File: rust/src/advanced/template_inheritance.rs

struct Pair<T, U> { first: T,  second: U, }
impl<T, U> Pair<T, U> {
  fn new(first: T, second: U) -> Self { Pair { first, second }  }
  fn set_value(&mut self, first: T, second: U) { self.first = first; self.second = second;  }
  fn get_first(&self) -> &T { &self.first  }
  fn get_second(&self) -> &U {  &self.second  }
}

// if we rename "pair:" with "super:" syntax is more devFriendly
struct LabeledPair<T, U> { pair: Pair<T, U>, label: String, }
impl<T, U> LabeledPair<T, U> {
  fn new(first: T, second: U, label: String) -> Self {
    LabeledPair { pair: Pair::new(first, second), label }
  }
  // ridefinizione obbligatoria getter oppure chiamata puntuale p.pair.get_second()
  fn get_first(&self) -> &T { self.pair.get_first()  }
  fn get_second(&self) -> &U {  self.pair.get_second()  }
  fn stampa_label(&self) { println!("Label: {}", self.label); }
}

pub fn inheritance_template_example() {
  let mut p = Pair::new(0.0, 0.0);
  p.set_value(10.0, 20.5);
  println!("First: {}, Second: {}", p.get_first(), p.get_second());

  // p.pair.get_second() return the same
  let p = LabeledPair::new(30, 40.5, "Example Label".to_string());
  println!("First: {}, Second: {}", p.get_first(), p.get_second()); 
  p.stampa_label();
}



